spring-integration-redis/src/main/java/org/springframework/integration/redis/inbound/RedisQueueMessageDrivenEndpoint.java

```diff
@@ -272,7 +272,7 @@ private void restart() {
 	protected void doStop() {
 		try {
 			this.active = false;
-			this.lifecycleCondition.await(Math.min(this.stopTimeout, this.receiveTimeout), TimeUnit.MICROSECONDS);
+			this.lifecycleCondition.await(Math.min(this.stopTimeout, this.receiveTimeout), TimeUnit.MILLISECONDS);
```

Comment:
Uh! Sorry, really have missed your point regarding `void stop(Runnable callback);`...
Not sure that we have something similar around, but the desired fix should not be so difficult to implement.